homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,442 @@
+"""Config flow for Apple TV integration."""
+import asyncio
+import logging
+from ipaddress import ip_address
+from random import randrange
+
+import voluptuous as vol
+
+from homeassistant import config_entries, core
+from homeassistant.const import (CONF_ADDRESS, CONF_NAME, CONF_PIN,
+                                 CONF_PROTOCOL, CONF_TYPE)
+from homeassistant.core import callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from pyatv import conf, const, convert, exceptions, pair, scan
+
+from .const import (CONF_CREDENTIALS, CONF_CREDENTIALS_AIRPLAY,
+                    CONF_CREDENTIALS_DMAP, CONF_CREDENTIALS_MRP,
+                    CONF_IDENTIFIER, CONF_START_OFF, DOMAIN)
+
+_LOGGER = logging.getLogger(__name__)
+
+INPUT_PIN_SCHEMA = vol.Schema({vol.Required(CONF_PIN, default=None): int})
+
+DEFAULT_START_OFF = False
+PROTOCOL_PRIORITY = [const.Protocol.MRP, const.Protocol.DMAP, const.Protocol.AirPlay]
+
+# Mapping between config entry format and pyatv
+CREDENTIAL_MAPPING = {
+    CONF_CREDENTIALS_MRP: const.Protocol.MRP.value,
+    CONF_CREDENTIALS_DMAP: const.Protocol.DMAP.value,
+    CONF_CREDENTIALS_AIRPLAY: const.Protocol.AirPlay.value,
+}
+
+
+async def device_scan(identifier, loop, cache=None):
+    def _filter_device(dev):
+        if identifier is None:
+            return True
+        if identifier == str(dev.address):
+            return True
+        if identifier == dev.name:
+            return True
+        return any([x.identifier == identifier for x in dev.services])
+
+    def _host_filter():
+        try:
+            return [ip_address(identifier)]
+        except ValueError:
+            return None
+
+    if cache:
+        matches = [atv for atv in cache if _filter_device(atv)]
+        if matches:
+            return cache, matches[0]
+
+    for hosts in [_host_filter(), None]:
+        scan_result = atvs = await scan(loop, timeout=3, hosts=hosts)
+        matches = [atv for atv in scan_result if _filter_device(atv)]
+
+        if matches:
+            return scan_result, matches[0]
+
+    return scan_result, None
+
+
+def is_valid_credentials(credentials):
+    return (
+        credentials.get(const.Protocol.MRP.value) is not None
+        or credentials.get(const.Protocol.DMAP.value) is not None
+    )
+
+
+@config_entries.HANDLERS.register(DOMAIN)
```

Comment:
This is the old way to register a domain. Just do class AppleTVFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):